zerver/lib/actions.py

```diff
@@ -2989,3 +2997,38 @@ def make_dict(row):
     streams.sort(key=lambda elt: elt["name"])
 
     return streams
+
+def do_process_attachment(message):
+
+    def extract_path_id(upload_string):
+        #We need to check if the string within the paranthesis contains a
+        #valid upload string
+        if re.search('[/\-]user[\-_]uploads[/\.-]', upload_string):
+            cleaned_path = re.sub('http://.*?/user_uploads/','', upload_string)
+            cleaned_path = re.sub('[\(\)]', '', cleaned_path)
+            return cleaned_path
+        else:
+            return None
+
+    #get all pairs of paranthesis
+    p = re.compile('\(.*?\)')
```

Comment:
I don't think using regular expressions here will produce all pairs of parens:
```
p.findall("(((a))(b))")
['(((a)', '(b)']
```

But that doesn't matter, because uploaded files don't need the parens.  I think we should just be matching for instances of the URLs containing the regular expression in `content_has_attachment` (or better, `upload_re` in zerver/lib/bugdown/__init__.py).

Structurally, I think extraction and tagging of attachments should probably be being done as a processing step inside the bugdown markdown processor.  Already, we have logic inside zerver/lib/bugdown/__init__.py for detecting uploaded files to set the title on links, and I think we should be able to reuse that logic for this purpose.

I think that change might be fairly complex to do for someone new to the Markdown codebase, so I think probably it'd be reasonable for now for you to fix this by just having this function look for instances of `/user_uploads/\d+)/[^/]*/([^/]*`, and extract the path from those. 

Does that make sense?  Happy to provide more explanation on anything that's confusing here.